applications/serial_lte_modem/src/twi/slm_at_twi.c

```diff
@@ -0,0 +1,404 @@
+
+	ret = i2c_read(slm_twi_dev[index], twi_data, (uint32_t)num_read, dev_addr);
+	if (ret < 0) {
+		LOG_ERR("Fail to read twi data");
+		return ret;
+	}
+
+	ret = slm_util_htoa(twi_data, num_read, twi_data_ascii, num_read*2);
+	if (ret > 0) {
+		sprintf(rsp_buf, "\r\n#XTWIR: %d\r\n", ret);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		rsp_send(twi_data_ascii, ret);
+		ret = 0;
+	} else {
+		LOG_ERR("hex convert error: %d", ret);
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int do_twi_write_read(uint16_t index, uint16_t dev_addr,
+			     const uint8_t *twi_data_ascii_w, uint16_t ascii_len,
+			     uint16_t num_read)
+{
+	int ret = -EINVAL;
+	uint8_t twi_data_w[ascii_len/2];
+	uint8_t twi_data_r[num_read], twi_data_ascii_r[num_read*2];
+
+	if (!slm_twi_dev[index]) {
+		LOG_ERR("TWI device is not opened");
+		return ret;
+	}
+
+	/* Decode hex string to hex array */
+	ret = slm_util_atoh(twi_data_ascii_w, ascii_len, twi_data_w, ascii_len/2);
+	if (ret < 0) {
+		LOG_ERR("Fail to decode hex string to hex array");
+		return ret;
+	}
+
+	ret = i2c_write_read(slm_twi_dev[index], dev_addr,
+			     twi_data_w, ret,
+			     twi_data_r, num_read);
+	if (ret < 0) {
+		LOG_ERR("Fail to write and read data at address: %hx", dev_addr);
+		return ret;
+	}
+
+	ret = slm_util_htoa(twi_data_r, num_read, twi_data_ascii_r, num_read*2);
+	if (ret > 0) {
+		sprintf(rsp_buf, "\r\n#XTWIWR: %d\r\n", ret);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		rsp_send(twi_data_ascii_r, ret);
+		ret = 0;
+	} else {
+		LOG_ERR("hex convert error: %d", ret);
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+/**@brief handle AT#XTWIOP commands
+ *  AT#XTWIOP=<index>
+ *  AT#XTWIOP? READ command not supported
+ *  AT#XTWIOP=?
+ */
+int handle_at_twi_open(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+
+	uint16_t idx;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		err = at_params_short_get(&at_param_list, 1, &idx);
+		if (err < 0) {
+			LOG_ERR("Fail to get twi index: %d", err);
+			return err;
+		}
+		err = do_twi_open(idx);
+		break;
+	case AT_CMD_TYPE_READ_COMMAND:
+		break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		sprintf(rsp_buf, "#XTWIOP: (0,1,2,3)\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		err = 0;
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+/**@brief handle AT#XTWIW commands
+ *  AT#XTWIW=<index>,<dev_addr>,<data>
+ *  AT#XTWIW? READ command not supported
+ *  AT#XTWIW=?
+ */
+int handle_at_twi_write(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	uint16_t index, dev_addr;
+	uint8_t twi_addr_ascii[TWI_ADDR_LEN + 1];
+	uint8_t twi_data_ascii[TWI_DATA_LEN + 1];
+	size_t ascii_len;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) != 4) {
+			LOG_ERR("Wrong input parameters");
+			return -EINVAL;
+		}
+		err = at_params_short_get(&at_param_list, 1, &index);
+		if (err < 0) {
+			LOG_ERR("Fail to get twi index: %d", err);
+			return err;
+		}
+		ascii_len = TWI_ADDR_LEN + 1;
```

Comment:
No, here since the null terminator is not expected, should not call our util but call at_cmd_parser to get the string directly

Reply:
The null terminator is expected. Then the address string is used as input of sscanf.